### PR TITLE
2480 - Add new command: migration-public:create

### PR DESCRIPTION
**Pull request by sorja**
image

### CODE REVIEW

File: src/tools/migrations/public/index.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import { Logger } from 'server/utils/logger'
+
+const client = DB
+let migrationSteps: Array<string>
+let previousMigrations: Array<string> = []
+const executedMigrations: Array<string> = []
+
+// TODO: Move to migration step / schema initialisation
+const tableDDL = `
+    create table if not exists public.migrations (
```
Comment (sorja) [inline src/tools/migrations/public/index.ts:18]: what should we call the db table? is `migrations` fine?

File: src/tools/migrations/public/index.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  previousMigrations = await client.map('select * from migrations', [], (row) => row.name)
+  migrationSteps = fs
+    .readdirSync(path.join(__dirname, `steps`))
+    .filter((file) => file !== 'template.ts' && file.endsWith('.ts') && !previousMigrations.includes(file))
+}
+
+const close = async () => {
+  // Note: Omitted REDIS related code. Add them here if needed.
+  await DB.$pool.end()
+}
```
Comment (sorja) [inline src/tools/migrations/public/index.ts:37]: for now i omit the redis related code, eg. connection closing etc as i don't think we need them in DDL manipulation
can be added as needed

File: src/tools/migrations/public/index.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+const exec = async () => {
+  await init()
+
+  // eslint-disable-next-line no-restricted-syntax
+  for await (const file of migrationSteps) {
+    try {
+      Logger.info(`Running migration ${file}`)
+      // eslint-disable-next-line @typescript-eslint/no-var-requires,global-require,import/no-dynamic-require
+      await require(`./steps/${file}`).default(client)
```
Comment (sorja) [inline src/tools/migrations/public/index.ts]: transcation has been removed here

it is also possible to omit passing client as we initialise the variable in the template or update template to take param

File: src/tools/migrations/public/index.ts (hunk 1 above)
Comment (minotogna) [reply src/tools/migrations/public/index.ts:18]: Sure 👍🏾

File: src/tools/migrations/public/index.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+const close = async () => {
+  // Note: Omitted REDIS related code. Add them here if needed.
+  await DB.$pool.end()
+}
+
+const exec = async () => {
+  await init()
+
+  // eslint-disable-next-line no-restricted-syntax
+  for await (const file of migrationSteps) {
```
Comment (minotogna) [inline src/tools/migrations/public/index.ts]: Is order by date 100% guaranteed?

File: src/tools/migrations/public/index.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+const close = async () => {
+  // Note: Omitted REDIS related code. Add them here if needed.
+  await DB.$pool.end()
+}
+
+const exec = async () => {
+  await init()
+
+  // eslint-disable-next-line no-restricted-syntax
```
Comment (minotogna) [inline src/tools/migrations/public/index.ts]: To avoid this, what about Promises.each ?

File: src/tools/migrations/public/index.ts (hunk 3 above)
Comment (minotogna) [reply src/tools/migrations/public/index.ts]: Client actually should not be passed , because ddl cannot be executed in transaction

**Review by minotogna — COMMENTED** [review]
Check deep scan as well

**Comment by sorja** [comment]
> Check deep scan as well 

Deepscan gives error on unused variable in the template file, what so you suggest?

**Comment by minotogna** [comment]
> > Check deep scan as well
> 
> Deepscan gives error on unused variable in the template file, what so you suggest?

remove the variable. we'll use DB directly in our migrations ?

File: src/tools/migrations/public/index.ts (hunk 4 above)
Comment (minotogna) [reply src/tools/migrations/public/index.ts]: @sorja is order guaranteed?

**Review by minotogna — APPROVED** [review]

File: src/tools/migrations/public/index.ts (hunk 4 above)
Comment (sorja) [reply src/tools/migrations/public/index.ts]: Should be fine!

image

image

**Pull request merged**